tests/test_messages_queue.cc

```diff
@@ -0,0 +1,482 @@
+      } else {
+        TEST_EQUAL(kTestMessage, FileManager::ReadFileAsString(full_file_path));
+        second_archived_file = full_file_path;
+        archive2_processed = true;
+      }
+      return true;  // Archives should be deleted by queue after successful
+                    // processing.
+    }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+    TEST_EQUAL(ProcessingResult::EProcessedSuccessfully, finish_task.get());
+    TEST_EQUAL(true, archive1_processed);
+    TEST_EQUAL(true, archive2_processed);
+    TEST_EQUAL("", FileManager::ReadFileAsString(archived_file));
+    TEST_EQUAL("", FileManager::ReadFileAsString(second_archived_file));
+  }
+}
+
+void Test_MessagesQueue_CreateArchiveOnSizeLimitHit() {
+  const std::string tmpdir = FileManager::GetDirectoryFromFilePath(GenerateTemporaryFileName());
+  CleanUpQueueFiles(tmpdir);
+  ScopedRemoveFile remover(tmpdir + alohalytics::kCurrentFileName);
+  MessagesQueue q;
+  q.SetStorageDirectory(tmpdir);
+
+  // Generate messages with total size enough for triggering archiving.
+  std::ofstream::pos_type size = 0;
+  auto const generator = [&q, &size](const std::string & message, std::ofstream::pos_type limit) {
+    std::ofstream::pos_type generated_size = 0;
+    while (generated_size < limit) {
+      q.PushMessage(message);
+      generated_size += message.size();
+    }
+    size += generated_size;
+  };
+  static const std::ofstream::pos_type limit = q.kMaxFileSizeInBytes / 2 + 100;
+  std::thread worker([&generator]() { generator(kTestWorkerMessage, limit); });
+  generator(kTestMessage, limit);
+  worker.join();
+
+  std::vector<std::ofstream::pos_type> file_sizes;
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&file_sizes](bool is_file, const std::string & full_file_path) {
+    TEST_EQUAL(true, is_file);
+    file_sizes.push_back(FileManager::ReadFileAsString(full_file_path).size());
+    return true;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::EProcessedSuccessfully, finish_task.get());
+  TEST_EQUAL(size_t(2), file_sizes.size());
+  TEST_EQUAL(size, file_sizes[0] + file_sizes[1]);
+  if (file_sizes[0] < q.kMaxFileSizeInBytes) {
+    TEST_EQUAL(true, file_sizes[1] > q.kMaxFileSizeInBytes);
+  } else {
+    TEST_EQUAL(true, file_sizes[0] > q.kMaxFileSizeInBytes);
+  }
+}
+
+void Test_MessagesQueue_HighLoadAndIntegrity() {
+  // TODO(AlexZ): This test can be improved by generating really a lot of data
+  // so many archives will be created. But it will make everything much more
+  // complex now.
+  const std::string tmpdir = FileManager::GetDirectoryFromFilePath(GenerateTemporaryFileName());
+  CleanUpQueueFiles(tmpdir);
+  ScopedRemoveFile remover(tmpdir + alohalytics::kCurrentFileName);
+  MessagesQueue q;
+  const int kMaxThreads = 300;
+  std::mt19937 gen(std::mt19937::default_seed);
+  std::uniform_int_distribution<> dis(1, std::numeric_limits<char>::max());
```

Comment:
If these are chars with no `'\0'`, then `[A..Z]` is just as good and less confusing.